### ISSUE TITLE
[4.2][Bug] Relationships don't get saved for subfields

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#4108]
I've tested this both against 4.2 and against https://github.com/Laravel-Backpack/CRUD/pull/4095 - neither work.

Here's what I've tried:
- in `InvoiceCrudController`, I wanted to add a relationship subfield for `InvoiceItems`; so I've:
   - defined the n-n polymorphic relationship `badgeable()` on the `InvoiceItem`
   - defined the reverse relationship on the `Badge`
   - added a new subfield to `InvoiceItems`:

```php
            [
                'name'       => 'badges',
                'type'       => 'text',
                'entity'     => 'badges',
            ],
```

So theoretically it should be... `Invoice` hasMany `InvoiceItem` morphToMany `Badge`.

But I'm greeted with a big fat error when I load the `Create` form:

![Screenshot 2022-01-21 at 18 26 44](https://user-images.githubusercontent.com/1032474/150563433-3fa4bd1e-0257-432a-8a9e-db0f2fbe4027.png)

I assume that it's because I haven't _completely_ defined the subfield. I remember that... if we use `relationship` as a subfield, we need to define every little bit of the relationship.

Then I've filled in manually the rest of the relationship definition:
```php
    CRUD::field('items')->subfields([
            // ...
            [
                'name'          => 'badges',
                'type'          => 'relationship',
                'entity'        => 'badges',
                'relation_type' => 'MorphToMany',
                'model'         => \App\Models\Badge::class,
                'attribute'     => 'name',
                'multiple'      => true,
                'pivot'         => true,
            ],
]);
```

Now a `select2_multiple` shows on for `Badges` just fine:

![Screenshot 2022-01-22 at 08 02 01](https://user-images.githubusercontent.com/1032474/150627030-64f90be2-27fe-4424-a0b7-c05023836c7a.png)

But when I try to save... boom:

![Screenshot 2022-01-22 at 07 55 58](https://user-images.githubusercontent.com/1032474/150626874-63bed157-bfa6-4bbf-8f82-e20c1b696ede.png)

**It looks like it doesn't care that I've defined the relationship attributes, it tries to save it as a regular db column.**

To sum up, in this example above (using a `MorphToMany` relationship as a subfield), I've encountered two problems:
- [x] the subfield did not infer the needed relationship info from the `entity`; fixed by https://github.com/Laravel-Backpack/CRUD/pull/4112 ;
- [x] the saving errorred (the field was taken as a normal field, not a relationship one);

_Originally posted by @tabacitu in https://github.com/Laravel-Backpack/CRUD/pull/4095#issuecomment-1019070027_

**Comment by tabacitu** [comment]
Ideas:

1. Maybe what we SHOULD do in this case (if that's indeed the problem) is give a clearer error message. Be more helpful to the developer. We could go as far as telling them "_When defining a subfield for a relationship, you have to manually define ALL relationship attributes: entity, model, relationship_type, x, y, z_". Otherwise... not even I know where to find all of that in the docs... haven't define those manually in ages.

2. Here's a crazy idea... can't we actually make it work inside subfields too? What if we had an extra relationship attribute, say `parentModel`, that would be used to detect the relationship:
    - by default, it would be the main model; nothing to define or overwrite;
    - but when using subfields inside a relationship, you could pass `parentModel => InvoiceItem::class` so that... when inferring the relationship attributes, it would look on the relationship on THAT model;

So:

```php
        CRUD::field('items')->subfields([
            // ...
            [
                'name'       => 'badges',
                'type'       => 'text',
                'entity'     => 'badges',
                'parentModel' => \App\Models\InvoiceItem::class,
            ],
        ])->reorder('order')->hint('<small class="float-right">Create/update/delete InvoiceItem entries over a <code>hasMany</code> relationship (1-n).</small>');
```

If `parentModel` is defined, addField would understand the `badges` relationship should be on `InvoiceItem`, NOT on `Invoice`. 

If possible, this would GREATLY improve the UX of devs using relationship as subfields. Instead of telling them "_define everything yourself_" we tell them "_define parentModel_".

Food for thought. Not sure it's possible. Not sure it's easy. Not sure if it's worth doing.

Update: Pedro said it's probably more difficult than expected here https://github.com/Laravel-Backpack/CRUD/pull/4095#issuecomment-1018782399

**Comment by tabacitu** [comment]
I've managed to do even better than that 😀  

In PR https://github.com/Laravel-Backpack/CRUD/pull/4112 I've managed to make relationship-attribute guessing in `subfields` work like `fields`. You don't have to give it any `entity` or `parentModel` or `baseModel` or anything... it'll know what model to look inside. You can override the `baseModel`... but you probably won't have to. It "_just works_". 

And I've achieved it in a pretty clean way, if I might say so 😅

![me_so_cool_cap](https://user-images.githubusercontent.com/1032474/150656892-b1cd53fa-6e21-4dcd-905d-623aaa2f711b.gif)

So if https://github.com/Laravel-Backpack/CRUD/pull/4112 gets merged, this problem is half-solved. The attributes will be inferred.

However... **n-n relationships still don't get saved, when they're in subfields.** I didn't fix that.

**Comment by tabacitu** [comment]
Basically, what was happening is that most relationships didn't work as subfields:
- [ ] hasOne (1-1) - doesn't work as subfield yet
- [x] belongsTo (n-1) - works as subfield thanks to https://github.com/Laravel-Backpack/CRUD/pull/4095
- [ ] hasMany (1-n) - doesn't work as subfield yet
- [ ] belongsToMany (n-n) - doesn't work as subfield yet
- [ ] morphOne (1-1) - doesn't work as subfield yet
- [ ] morphMany (1-n) - doesn't work as subfield yet
- [ ] morphToMany (n-n) - doesn't work as subfield yet

 @pxpm is on this, and will submit one or more PRs to make the rest of them work. When he does so, I will update the above, and the [subfields breakdown](https://github.com/Laravel-Backpack/CRUD/issues/4109#issuecomment-1019323850).

**Comment by tabacitu** [comment]
Just tested all relationships inside two out of three scenarios - they work GREAT 🎉

I'm going to close this, and open a different issue for us to check/add support for Models being used as pivot fields - https://github.com/Laravel-Backpack/CRUD/issues/4142
